File: backend-hono/src/lib/srt.ts
/**
 * transcript の SRT 形式検証（VideoUpdateSerializer.validate_transcript と一致）。
 * 空/空白のみは検証をスキップ。各ブロックは index 行(整数)/timestamp 行/本文 の 3 行以上。
 */
const SRT_TIMESTAMP_RE =
  /^\d{2}:\d{2}:\d{2},\d{3}\s+-->\s+\d{2}:\d{2}:\d{2},\d{3}$/;

export const INVALID_SRT_MESSAGE = "Transcript must be in valid SRT format.";

/** 不正なら INVALID_SRT_MESSAGE、妥当なら null。 */
export function validateTranscriptSrt(value: string): string | null {
  if (!value || value.trim() === "") return null;

  const blocks = value
    .split("\n\n")
    .map((b) => b.trim())
    .filter((b) => b);
  for (const block of blocks) {
    const lines = block.split("\n");
    if (lines.length < 3) return INVALID_SRT_MESSAGE;
    // int(lines[0].strip())（符号付き整数のみ）
    if (!/^[+-]?\d+$/.test(lines[0].trim())) return INVALID_SRT_MESSAGE;
    if (!SRT_TIMESTAMP_RE.test(lines[1].trim())) return INVALID_SRT_MESSAGE;
  }
  return null;
}

/** TimestampConverter.parse_timestamp 相当（HH:MM:SS,mmm / HH:MM:SS / ドット区切り）。 */
export function parseSrtTimestamp(timestamp: string): number {
  const normalized = timestamp.replace(/\./g, ",");
  const parts = normalized.split(",");
  const timeParts = parts[0]!.split(":").map((p) => Number.parseInt(p, 10));
  let hours: number;
  let minutes: number;
  let secondsPart: number;
  if (timeParts.length === 3) {
    [hours, minutes, secondsPart] = timeParts as [number, number, number];
  } else if (timeParts.length === 2) {
    hours = 0;
    [minutes, secondsPart] = timeParts as [number, number];
  } else {
    throw new Error(`Invalid timestamp: ${timestamp}`);
  }
  let seconds = hours * 3600 + minutes * 60 + secondsPart;
  if (parts.length > 1) seconds += Number.parseInt(parts[1]!, 10) / 1000;
  return seconds;
}

export type SrtScene = {
  index: number | null;
  start_time: string;
  end_time: string;
  start_sec: number;
  end_sec: number;
  text: string;
};

/** SubtitleParser.parse_srt_scenes 相当（L0 Retrieve 用）。 */
export function parseSrtScenes(srtString: string): SrtScene[] {
  const content = srtString.trim();
  if (!content) return [];
  const blocks = content
    .split("\n\n")
    .map((b) => b.trim())
    .filter((b) => b);
  const scenes: SrtScene[] = [];
  for (const block of blocks) {
    const lines = block.split("\n");
    if (lines.length < 3) continue;
    let index: number | null = null;
    try {
      index = Number.parseInt(lines[0]!.trim(), 10);
      if (Number.isNaN(index)) index = null;
    } catch {
      index = null;
    }
    const timing = lines[1]!.trim();
    if (!timing.includes("-->")) continue;
    const [startStr, endStr] = timing.split("-->").map((t) => t.trim());
    if (!startStr || !endStr) continue;
    const text = lines
      .slice(2)
      .map((l) => l.trim())
      .filter((l) => l)
      .join(" ");
    scenes.push({
      index,
      start_time: startStr,
      end_time: endStr,
      start_sec: parseSrtTimestamp(startStr),
      end_sec: parseSrtTimestamp(endStr),
      text,
    });
  }
  return scenes;
}

File: apps/worker/worker_python/pipeline/srt.py
"""Shared SRT parsing and formatting for transcription, scenes and indexing."""

from __future__ import annotations

import re
from collections.abc import Iterator
from dataclasses import dataclass


@dataclass
class SrtScene:
    index: int | None
    start_time: str
    end_time: str
    start_sec: float
    end_sec: float
    text: str


def parse_srt_timestamp(timestamp: str) -> float:
    normalized = timestamp.replace(".", ",")
    parts = normalized.split(",")
    time_parts = [int(p) for p in parts[0].split(":")]
    if len(time_parts) == 3:
        hours, minutes, seconds_part = time_parts
    elif len(time_parts) == 2:
        hours = 0
        minutes, seconds_part = time_parts
    else:
        raise ValueError(f"Invalid timestamp: {timestamp}")
    seconds = hours * 3600 + minutes * 60 + seconds_part
    if len(parts) > 1:
        seconds += int(parts[1]) / 1000.0
    return float(seconds)


def format_srt_time(seconds: float) -> str:
    total_millis = round(max(seconds, 0.0) * 1000)
    hours, remainder = divmod(total_millis, 3_600_000)
    minutes, remainder = divmod(remainder, 60_000)
    whole, millis = divmod(remainder, 1000)
    return f"{hours:02d}:{minutes:02d}:{whole:02d},{millis:03d}"


def create_srt_from_whisper_segments(segments: list[dict]) -> str:
    lines: list[str] = []
    for i, seg in enumerate(segments, start=1):
        text = str(seg.get("text", "")).strip()
        if not text:
            continue
        start = float(seg.get("start", 0))
        end = float(seg.get("end", start))
        lines.append(str(i))
        lines.append(f"{format_srt_time(start)} --> {format_srt_time(end)}")
        lines.append(text)
        lines.append("")
    return "\n".join(lines)


def parse_srt_scenes(srt_string: str) -> list[SrtScene]:
    return list(iter_srt_scenes(srt_string))


def _iter_srt_blocks(srt_string: str) -> Iterator[str]:
    content = srt_string.replace("\r\n", "\n").replace("\r", "\n").strip()
    start = 0
    for separator in re.finditer(r"\n[ \t]*\n", content):
        yield content[start : separator.start()]
        start = separator.end()
    yield content[start:]


def iter_srt_scenes(srt_string: str) -> Iterator[SrtScene]:
    """Yield valid scenes so bounded consumers can stop without parsing the rest."""
    for block in _iter_srt_blocks(srt_string):
        block = block.strip()
        if not block:
            continue
        lines = block.split("\n")
        if len(lines) < 3:
            continue
        index: int | None
        try:
            index = int(lines[0].strip())
        except ValueError:
            index = None
        timing = lines[1].strip()
        if "-->" not in timing:
            continue
        start_str, end_str = [t.strip() for t in timing.split("-->", 1)]
        text = "\n".join(lines[2:]).strip()
        if not text:
            continue
        try:
            start_sec = parse_srt_timestamp(start_str)
            end_sec = parse_srt_timestamp(end_str)
        except ValueError:
            continue
        yield SrtScene(
            index=index,
            start_time=start_str,
            end_time=end_str,
            start_sec=start_sec,
            end_sec=end_sec,
            text=text,
        )
